docs/source/receive-adapter.md

```diff
@@ -0,0 +1,156 @@
+# Receive adapter
+
+A _receive adapter_ is a data-plane component which either
+
+- produces [CloudEvents](https://github.com/cloudevents/spec),
+- or imports events and convert them to
+  [CloudEvents](https://github.com/cloudevents/spec),
+
+and forwards those events to one or several consumers (aka _sink_),
+depending upon whether it supports multiple resources (see below for more information).
+
+## Implementating receive adapter
+
+Knative Eventing provides a library (refer later as *the library*) for bootstrapping receive adapter implementations.
+
+The main code should be something like this:
+
+```go
+package main
+
+import (
+  "knative.dev/eventing/pkg/adapter/youradapter"
+  "knative.dev/eventing/pkg/adapter/v2"
+)
+
+func main() {
+  adapter.Main("pingsource",
+    youradapter.NewEnvConfig,
+    youradapter.NewAdapter)
+}
+```
+
+`adapter.Main` is responsible for:
+
+- parsing the command flags and initializing the environment configuration
+returned by `youradapter.NewEnvConfig`,
+- setting up the logger,
+- setting up the profiler server (when enabled),
+- setting up the metrics server,
+- setting up tracing,
+- setting up a ConfigMap watcher (when enabled),
+- setting up a CloudEvent client, targeting a single sink, for single resource adapter,
+or otherwise targeting no sink,
+- and finally running *your adapter* in leader election mode (when enabled) for being
+highly-available and scalable.
+
+## Push vs pull models
+
+The push model describes the processing mode under which the receive adapter
+reacts to incoming requests or events. The pull model is the opposite, the
+receive adapter does not wait for incoming requests or events but instead it fetches,
+or generates events.
+
+The receive adapter library can be used to implement either type of receive adapters.
+For pull-based receive adapters, the library provides a mechanism for making your adapter
+highly-available and scalable.
+
+## Handling multiple resources (aka multi-tenancy)
+
+### Overview
+
+Single-resource receive adapters handle one source instance (CR) at a time, i.e.
```

Comment:
I believe the correct term is `CO` instead of `CR` in the kubernetes documentation?

Reply:
this page talks about custom resource: https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/

I'll expand it to make it clear that's a custom resource.